client/client-lib/src/api.rs

```diff
@@ -76,15 +79,22 @@ impl<'a> dyn FederationApi + 'a {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 /// Mint API client that will try to run queries against all `members` expecting equal
 /// results from at least `min_eq_results` of them. Members that return differing results are
 /// returned as a member faults list.
 pub struct WsFederationApi<C> {
-    clients: Vec<(PeerId, C)>,
+    members: Vec<FederationMember<C>>,
     max_evil: usize,
 }
 
+#[derive(Debug)]
+struct FederationMember<C> {
+    url: String,
+    peer_id: PeerId,
+    client: RwLock<Option<C>>,
```

Comment:
I'm a bit scared of having that many locks that are being acquired and released in an unspecified manner. But after thinking about it for a while I understand why and don't have a better alternative. I think it's probably ok, we'll just have to reason about lock acquisition and release very carefully.